### PR TITLE
TS-4388: Fix TSHttpTxnParentProxySet

### PR DESCRIPTION
**Pull request by jpeach**
This PR fixes ``TSHttpTxnParentProxySet``.

The first two commits are minor cleanups of cruft. The third one fixes the API tests so that they don't interfere with each other when they register multiple global hooks. The final one actually fixes the ``TSHttpTxnParentProxySet`` API, primarily by restricting access to private ``ParentResult`` data and providing accessors that enforce the correct policy.

I've tested that this works correctly with the tests and with my plugin. I don't have any way to test that I haven't broken the various parent retry features.

### CODE REVIEW

**Comment by jpeach** [comment]
Ping @zwoop @PSUdaemon @jrushford

**Comment by zwoop** [comment]
I'm not sure I'm in love with

    result->result

That seems pretty redundant. If you are going to bike shed this, maybe result->rc, or result->result_code or result->state ?

**Comment by JamesPeach** [comment]
The 99% case in the code is ``parent_result.result`` which didn't seem too bad to me. I hated, hated, hated ``parent_result.r`` ;)

File: proxy/http/HttpConfig.cc [hunk 1]
```diff
@@ -888,8 +888,6 @@ HttpConfig::startup()
   HttpEstablishStaticConfigLongLong(c.origin_min_keep_alive_connections, "proxy.config.http.origin_min_keep_alive_connections");
   HttpEstablishStaticConfigLongLong(c.oride.attach_server_session_to_client, "proxy.config.http.attach_server_session_to_client");
 
-  HttpEstablishStaticConfigByte(c.parent_proxy_routing_enable, "proxy.config.http.parent_proxy_routing_enable");
-
   // Wank me.
```
Comment (PSUdaemon) [inline proxy/http/HttpConfig.cc]: Didn't think this cruft was worth removing? ;-)

**Comment by PSUdaemon** [comment]
The 4th commit is so much refactoring with no change in behavior that it's hard to follow what the actual fix is. I agree that `r` is a poor variable name, and a lot of the other changes do make the code clearer. Would be nice to separate them out as you did in the other commits.

**Comment by jpeach** [comment]
The actual fix is to avoid dereferencing the ``rec`` pointer when it has the API sentinel value. I could have just make a small fix to the code path that I happened to hit, but I wanted to make it robust in all cases. I could have separated out each usage of ``rec`` into a separate commit (which was actually how I worked this). It would be a long series.

**Comment by PSUdaemon** [comment]
I like the robustness, it's just hard to tell that this commit is anything more than a refactor. I'd just separate the small fix out and leave the rest as one patch.

**Comment by jpeach** [comment]
I looked at the diff again, and though there are a few changes here, they are pretty obvious in a side-by-side diff. Basically every place that follows ``parent_record.rec`` unchecked is a bug.

**Comment by zwoop** [comment]
I agree with Phil, since the refactoring (variable name changes) is so big, it would be much nicer to have this as two separate PRs.

**Comment by jpeach** [comment]
OK I broke out some refactoring into 2 additional commits.

**Comment by zwoop** [comment]
+1 on the refactoring / commits.

**Comment by jpeach** [comment]
Merged.
